app/main/views/users.py

```diff
@@ -271,71 +281,123 @@ def export_users_for_framework(framework_slug):
 
     suppliers_with_a_complete_service = frozenset(framework.get_supplier_ids_for_completed_service())
 
-    supplier_id_published_service_count = dict(db.session.query(
-        Service.supplier_id,
-        func.count(Service.id)
+    users = db.session.query(
+        User.supplier_id, User.name, User.email_address, User.user_research_opted_in,
+        SupplierFramework.framework_id, SupplierFramework.supplier_id,
+    ).filter(
+        SupplierFramework.supplier_id == User.supplier_id
     ).filter(
-        Service.status == 'published',
-        Service.framework_id == framework.id
-    ).group_by(
-        Service.supplier_id
-    ).all())
+        SupplierFramework.framework_id == framework.id
+    ).filter(
+        User.active.is_(True)
+    ).all()
+
+    users_for_each_supplier = {}
+    for u in users:
+        user_obj = {
+            'email_address': u.email_address,
+            'name': u.name,
+            'user_research_opted_in': u.user_research_opted_in,
+        }
+        # If supplier id key already exists, append the user to that supplier's list
+        users_for_each_supplier.setdefault(u.supplier_id, []).append(user_obj)
+
+    lot_slugs_by_id = {lot.id: lot.slug for lot in framework.lots}
+
+    # Creates a dictionary of dictionaries, with published service counts per lot per supplier.
+    # To do this we use dictionary comprehension, SQL query and itertools' 'groupby' function.
+    # Grouping by supplier id is more efficient as this is what we will be iterating over later.
+    service_counts_by_lot_by_supplier = {
+        supplier_id: {row[1]: row[2] for row in rows}
+        for supplier_id, rows in groupby(
```

Comment:
Nested comprehensions like this are hard to read. This would look neated assigning the result of the query to a variable.
https://google.github.io/styleguide/pyguide.html#List_Comprehensions

Reply:
At a glance it looks like you're starting a for loop here with:
```
for supplier_id, rows in groupby(
```
It's only when we get to line 325 that we realise it's a comprehension and have to navigate back to line 310 to start reading the whole thing over again.

Reply:
I don't think it at all looks like you're starting a for loop there - the simple fact that it's indented and inside braces means that it *can't* be a for loop.

As for nested comprehensions, breaking out to an additional variable requires the user to do more mental variable tracking, makes the reader wonder whether the intermediate has managed to be mutated (as it has leaked a reference to its inner working) and adds yet another thing to assign a name to (and, though in an ideal world we're all great at naming things, in reality "naming things" is one of the "hard problems")

Reply:
but really, assigning the query to a variable isn't a huge change, not a big deal, won't fight tooth and nail, just think it's harder to read, making me go and link up the innards of the comprehension with where it's being used. Scrolly-uppy, scrolly-downy. Scrolly-uppy...

...and then wonder about the possibilities of whether I make a change to it whether something else is affected by it. Inside the comprehension it is *encapsulated*.